src/services/userValidation.ts

```diff
@@ -1,8 +1,63 @@
-import { validationResult } from "express-validator";
-import type { Request } from "express";
-import { validators } from "../utils/registerValidators";
+import * as zod from "zod";
+import { findUserBy } from "./databaseServices/findUserBy";
 
-export const validateUserData = async (req: Request) => {
-  await Promise.all(validators.map((validator) => validator.run(req)));
-  return validationResult(req);
+const registerSchema = zod.object({
+  username: zod
+    .string()
+    .min(8, { message: "Username must be at least 8 chars long" }),
+  password: zod
+    .string()
+    .min(8, { message: "Password must be at least 8 chars long" }),
+  email: zod.string().email({ message: "Invalid email address" }),
+});
+
+const loginSchema = zod.object({
+  username: zod
+    .string()
+    .min(8, { message: "Username must be at least 8 chars long" }),
+  password: zod
+    .string()
+    .min(8, { message: "Password must be at least 8 chars long" }),
+});
+
+export const validateRegister = async ({
+  username,
+  password,
+  email,
+}: registerData) => {
+  try {
+    registerSchema.parse({ username, password, email });
+    const usernameFound = await findUserBy({ username });
+    const emailFound = await findUserBy({ email });
+    if (usernameFound || emailFound) {
+      throw [{ message: "Username or email already taken" }];
+    }
+    return true;
```

Comment:
to me this function has an AWFUL API.

To check if the validation succedeed I'd need to compare to true explicitly: `validateRegister(...) === true`